data_model/src/predicate.rs

```diff
@@ -2,54 +2,155 @@
 
 #[cfg(not(feature = "std"))]
 use alloc::vec;
+use core::{fmt::Display, ops::Not};
 
 use super::*;
 use crate::{IdBox, Name, Value};
 
+/// Struct representing sequence of non-zero length
+#[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, IntoSchema)]
+pub struct NonEmpty<T> {
+    head: Box<T>,
+    tail: Vec<T>,
+}
+
+macro_rules! nonempty {
+    ($h:expr, $( $t:expr ),*) => {{
+        NonEmpty {
+            head: Box::new($h),
+            tail: vec![$($t),*],
+        }
+    }};
+}
+
 /// Predicate combinator enum.
 #[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, IntoSchema)]
-pub enum PredicateBox {
-    /// Logically `&&` the results of applying the two predicates.
-    And(Vec<PredicateBox>),
-    /// Logically `||` the results of applying the two predicates.
-    Or(Vec<PredicateBox>),
+// Ideally we would enforce `P: PredicateTrait<Input>` here, but I
+// couldn't find a way to do it without polluting everything
+// downstream with explicit lifetimes, since we would need to
+// store PhantomData<Input> here, and `Input`s are most often
+// references (e.g. &Value).
+pub enum GenericPredicateBox<P> {
+    /// Logically `&&` the results of applying the predicates.
+    And(NonEmpty<Self>),
+    /// Logically `||` the results of applying the predicates.
+    Or(NonEmpty<Self>),
     /// Negate the result of applying the predicate.
-    Not(Box<PredicateBox>),
+    Not(Box<Self>),
     /// The raw predicate that must be applied.
-    Raw(value::Predicate),
+    Raw(P),
 }
 
-impl PredicateBox {
+impl<P> Display for GenericPredicateBox<P>
+where
+    P: Display,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            GenericPredicateBox::And(predicates) => {
+                write!(f, "OR(")?;
```

Comment:
Is this correct? we match `And` and print `OR`.

Reply:
Incorrect indeed, thanks